compiler/rustc_pattern_analysis/src/usefulness.rs

```diff
@@ -1440,10 +1441,17 @@ fn compute_exhaustiveness_and_usefulness<'a, 'p, Cx: TypeCx>(
     // We treat match scrutinees of type `!` or `EmptyEnum` differently.
     let is_toplevel_exception =
         is_top_level && matches!(ctors_for_ty, ConstructorSet::NoConstructors);
-    // Whether empty patterns can be omitted for exhaustiveness.
-    let can_omit_empty_arms = is_toplevel_exception || mcx.tycx.is_exhaustive_patterns_feature_on();
-    // Whether empty patterns are counted as useful or not.
-    let empty_arms_are_unreachable = place_validity.is_known_valid() && can_omit_empty_arms;
+    // Whether empty patterns are counted as useful or not. We only warn an empty arm unreachable if
+    // it is guaranteed unreachable by the opsem (i.e. if the place is `known_valid`).
+    let empty_arms_are_unreachable = place_validity.is_known_valid()
```

Comment:
it was hard to understand what was going on here until i noticed that:

1. these bools flipped order
2. the old `can_omit_empty_arms` was "inlined" into `empty_arms_are_unreachable`
3. the new `can_omit_empty_arms` is *not* enabled for the new min feature gate.

Reply:
I tried so hard to make it legible :') The new `can_omit_empty_arms` *is* enabled with the feature gate, but only exactly when `empty_arms_are_unreachable` is true, i.e. the place must be `known_valid`.

I tried to present it as "the base case is `can_omit_empty_arms == empty_arms_are_unreachable`, on top of which there are two exceptions" but I guess that wasn't clear

Reply:
Ah, right -- `empty_arms_are_unreachable ||` in the new `can_omit_empty_arms`.